meadow-client/src/Semantics.purs

```diff
@@ -0,0 +1,1067 @@
+data WIdChoice = WIdChoice IdChoice
+derive instance eqWIdChoice :: Eq WIdChoice
+derive instance ordWIdChoice :: Ord WIdChoice
+
+type IdOracle = BigInt
+
+type LetLabel = BigInt
+
+showIdChoice :: IdChoice -> String
+showIdChoice { choice : choice, person : person } =
+      joinWith "" ["(", toString choice, ", ", toString person, ")"]
+
+data Value = CurrentBlock |
+             Committed IdCommit |
+             Constant BigInt |
+             NegValue Value |
+             AddValue Value Value |
+             SubValue Value Value |
+             MulValue Value Value |
+             DivValue Value Value Value |
+ --          dividend-^ divisor-^ ^-default value (in case divisor is zero)
+             ModValue Value Value Value |
+ --          dividend-^ divisor-^ ^-default value (in case divisor is zero)
+             ValueFromChoice IdChoice Value |
+ --    default value if not available --^
+             ValueFromOracle IdOracle Value
+ --    default value if not available --^
+derive instance eqValue :: Eq Value  
+derive instance ordValue :: Ord Value  
+
+showValue :: Value -> String
+showValue CurrentBlock = "CurrentBlock"
+showValue (Committed idCommit) = joinWith "" ["Committed ", toString idCommit]
+showValue (Constant bigInt) = joinWith "" ["Constant ", toString bigInt] 
+showValue (NegValue value1) = joinWith "" ["NegValue (", showValue value1, ")"] 
+showValue (AddValue value1 value2) = joinWith "" ["AddValue (", showValue value1,
+                                                         ") (", showValue value2, ")"] 
+showValue (SubValue value1 value2) = joinWith "" ["SubValue (", showValue value1,
+                                                         ") (", showValue value2, ")"]
+showValue (MulValue value1 value2) = joinWith "" ["MulValue (", showValue value1,
+                                                         ") (", showValue value2, ")"]
+showValue (DivValue value1 value2 value3) = joinWith "" ["DivValue (", showValue value1,
+                                                                ") (", showValue value2,
+                                                                ") (", showValue value3, ")"]
+showValue (ModValue value1 value2 value3) = joinWith "" ["ModValue (", showValue value1,
+                                                                ") (", showValue value2,
+                                                                ") (", showValue value3, ")"]
+showValue (ValueFromChoice idChoice value) = joinWith "" ["ValueFromChoice ",
+                                                          showIdChoice idChoice,
+                                                          " (", showValue value, ")"]
+showValue (ValueFromOracle idOracle value) = joinWith "" ["ValueFromOracle ",
+                                                          toString idOracle,
+                                                          " (", showValue value, ")"]
```

Comment:
Hi, I'm from the Church of Recursion Schemes. Can I show you some pamphlets?

Reply:
Recur... what?